inlong-sort/sort-flink/sort-flink-v1.15/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/FlinkTubeMQProducer.java

```diff
@@ -0,0 +1,178 @@
+        checkNotNull(masterAddress, "The master address must not be null.");
+        checkNotNull(serializationSchema, "The serialization schema must not be null.");
+        checkNotNull(tidSet, "The tid set must not be null.");
+        checkNotNull(configuration, "The configuration must not be null.");
+
+        this.topic = topic;
+        this.masterAddress = masterAddress;
+        this.serializationSchema = serializationSchema;
+        this.tidSet = tidSet;
+        this.maxRetries = configuration.getInteger(TubeMQOptions.MAX_RETRIES);
+    }
+
+    @Override
+    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) {
+        // Nothing to do.
+    }
+
+    @Override
+    public void initializeState(FunctionInitializationContext functionInitializationContext) {
+        // Nothing to do.
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+
+        TubeClientConfig tubeClientConfig = new TubeClientConfig(masterAddress);
+        this.sessionFactory = new TubeSingleSessionFactory(tubeClientConfig);
+        this.producer = sessionFactory.createProducer();
+        HashSet<String> hashSet = new HashSet<>();
+        hashSet.add(topic);
+        producer.publish(hashSet);
+    }
+
+    @Override
+    public void invoke(T in, SinkFunction.Context context) throws Exception {
+
+        int retries = 0;
+        Exception exception = null;
+
+        while (maxRetries <= 0 || retries < maxRetries) {
+
+            try {
+                byte[] body = serializationSchema.serialize(in);
+                Message message = new Message(topic, body);
+                // if (!tidSet.isEmpty()) {
+                // SimpleDateFormat sdf = new SimpleDateFormat(SYSTEM_HEADER_TIME_FORMAT);
+                // long currTimeMillis = System.currentTimeMillis();
+                // message.putSystemHeader(tidSet.toString(), sdf.format(new Date(currTimeMillis)));
+                // }
+                MessageSentResult sendResult = producer.sendMessage(message);
+                if (sendResult.isSuccess()) {
+                    return;
+                } else {
+                    LOG.warn("Send msg fail, error code: {}, error message: {}",
+                            sendResult.getErrCode(), sendResult.getErrMsg());
+                }
+            } catch (Exception e) {
+                LOG.warn("Could not properly send the message to hippo "
+                        + "(retries: {}).", retries, e);
+
+                retries++;
+                exception = ExceptionUtils.firstOrSuppressed(e, exception);
+            }
+        }
+
+        throw new IOException("Could not properly send the message to hippo.", exception);
+    }
+
+    @Override
+    public void close() throws Exception {
+
+        try {
+            if (producer != null) {
+                producer.shutdown();
+                producer = null;
```

Comment:
There is no need to assign null to the producer.
It will throw NPE rather than a real Exception like ProducerAlreadyClosedException